### PR TITLE
Keystore vault

### LINKED ISSUE #17252: Add Keycloak Keystore Vault implementation
Keyloak should be able to use Java keystore as a vault. It will be implemented as an another implementation of the Vault SPI.

More information can be found here [1].

[1] https://github.com/keycloak/keycloak/discussions/19281

### PR DESCRIPTION
**Pull request by Pepo48**
KeystoreVault SPI

* added KeystoreVault - a Vault SPI implementation (#19281)

Closes #17252

Signed-off-by: Peter Zaoral <pzaoral@redhat.com>

### CODE REVIEW

**Review by pedroigor — DISMISSED** [review]

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/VaultKeystorePropertyMappers.java
Comment (pedroigor) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/VaultKeystorePropertyMappers.java:25]: I think we should use it to resolve the type depending on how FIPS is enabled too. See https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HttpPropertyMappers.java#L146.

Comment (Pepo48) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/VaultKeystorePropertyMappers.java:32]: Thanks for the review, @pedroigor. Is this how you meant it?
https://github.com/keycloak/keycloak/pull/19644/commits/2488d32d7692be5eb330d5256b36ce9e381dbf5d#diff-08975c153576d59b1d82e056efd2ba2cc67d82dd1e330c2fc78f4b91ab664d55R34-R56

**Comment by Pepo48** [comment]
There are still some failures, mainly related to `HelpCommandDistTest` - I will update the templates tomorrow, that should fix it.

File: docs/guides/server/vault.adoc
Comment (abstractj) [inline docs/guides/server/vault.adoc:67]: Before having this change in KC 22. Would be possible to update/replace the algorithm provided by Keycloak Vault?

MD5 and DES combination cannot easily be broken but brute-forced, considering that we are aiming to enhance security, I'd suggest replacing it by PBEWithHmacSHA256AndAES_128, or PBEWithHmacSHA256AndAES_256 if possible.

@JoshuaMulliken could you please review? The proposal is based on https://github.com/keycloak/keycloak/discussions/19281

Comment (JoshuaMulliken) [reply docs/guides/server/vault.adoc:67]: I agree and we should 100% utilize the aes256 variant as it will set us up for success when the Feds post quantum encryption requirements come into affect

Comment (abstractj) [reply docs/guides/server/vault.adoc:67]: @vmuzikar @Pepo48 can we change the instructions to something like the snipped below?

```
keytool -importpass -alias test_alias -keystore keystore.p12 -keyalg PBEWithHmacSHA256AndAES_256 -storepass keystorepassword
```

Or it does not matter what we do, the keystore will always use the default?

Comment (Pepo48) [reply docs/guides/server/vault.adoc:67]: @abstractj I actually had the very same idea while testing it, but based on my experience this does not change Secret Key Entry algorithm. Here's a quick proof:
image

File: services/src/test/java/org/keycloak/vault/KeystoreVaultProviderTest.java
Comment (vmuzikar) [inline services/src/test/java/org/keycloak/vault/KeystoreVaultProviderTest.java:18]: This is a unit test. Would it make sense to also add a Quarkus dist test for it? WDYT @Pepo48 @pedroigor

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@Pepo48 Thank you, very nice work!

Added a few comments. On top of that, could you please add `org.keycloak.vault.FilesKeystoreVaultProviderFactory` here:
https://github.com/keycloak/keycloak/blob/cda0c9dce0accbb38f1efee23fd7430a0bac5deb/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java#L165-L178
This is required so that only the Quarkus version of the factory is loaded.

**Review by pedroigor — DISMISSED** [review]
@Pepo48 Looks like the remaining comments are outdated. From my side, it looks good.

**Review by abstractj — DISMISSED** [review]

**Comment by abstractj** [comment]
@vmuzikar @Pepo48 before merging it, I'd suggest squashing those commits. I approved it, regardless.

**Comment by abstractj** [comment]
@Pepo48 we need to resolve those conflicts.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakKeystoreVaultFipsModeTest.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakKeystoreVaultFipsModeTest.java:40]: I think we also need to use a  different keystore file for this test.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakPlaintextVaultTest.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakPlaintextVaultTest.java:94]: I might be missing something but why is plaintext test using a keystore? Is that correct? In any case, it doesn't seem correct to me that we expect exceptions for each Vault provider.

File: services/src/main/java/org/keycloak/vault/FilesKeystoreVaultProvider.java
Comment (abstractj) [inline services/src/main/java/org/keycloak/vault/FilesKeystoreVaultProvider.java:68]: @Pepo48 it would be nice to mitigate _Path Traversal_ before we merge those changes by escaping or validating the user input when _keystorePath_ is provided. Otherwise, we may introduce a new vulnerability.

Please see:
- https://owasp.org/www-community/attacks/Path_Traversal
- https://codeql.github.com/codeql-query-help/java/java-path-injection/

**Review by abstractj — DISMISSED** [review]
@Pepo48 thanks for looking into the issues for path traversal. It appears that we were able to mitigate it.

**Comment by abstractj** [comment]
@Pepo48 please squash those commits into a single one to reduce confusion, unless it is strictly necessary to have several commits.

**Comment by Pepo48** [comment]
> @Pepo48 please squash those commits into a single one to reduce confusion, unless it is strictly necessary to have several commits.

@abstractj I am going to do it at the end, as always. There are still some uncertainties around FIPS, which prevents me from wrapping this up, see the recent news here: https://github.com/quarkusio/quarkus/issues/33314

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java:40]: I think we should rather make the `KeycloakVaultTest` abstract and create two derived classes – `KeycloakPlaintextVaultTest` and `KeycloakKeystoreVaultTest` (we already have that one).

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java:86]: I think  we should remove this now that the FIPS issues are resolved.

**Review by vmuzikar — CHANGES_REQUESTED** [review]

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java
Comment (Pepo48) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java:86]: This has nothing to do with FIPS though. It tests invalid entry in the vault.

Comment (vmuzikar) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java:86]: Then we should probably have some switch or something to expect the exception only when relevant.

File: docs/guides/server/vault.adoc
Comment (Pepo48) [reply docs/guides/server/vault.adoc]: no, it wasn't intentional; thanks!

File: docs/guides/server/vault.adoc
Comment (Pepo48) [reply docs/guides/server/vault.adoc]: done

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/FipsDistTest.java
Comment (Pepo48) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/FipsDistTest.java]: done - I reverted all the changes with regards to FipsDistTest

File: services/src/main/java/org/keycloak/vault/FilesKeystoreVaultProvider.java
Comment (Pepo48) [reply services/src/main/java/org/keycloak/vault/FilesKeystoreVaultProvider.java]: done

File: services/src/main/java/org/keycloak/vault/FilesKeystoreVaultProvider.java
Comment (Pepo48) [reply services/src/main/java/org/keycloak/vault/FilesKeystoreVaultProvider.java]: correct, we don't, I missed during the fips cleanup. Thanks.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakKeystoreVaultFipsModeTest.java
Comment (Pepo48) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakKeystoreVaultFipsModeTest.java]: For some reason I can't add a comment under the `KeystoreVaultProviderTest.java`, regardless I added two tests under the `ConfigurationTest.java`.  Please check the changes within `testEnvVarAvailableFromPropertyNames` and `testPropertyNamesFromConfig` tests.

File: testsuite/integration-arquillian/tests/base/testsuites/fips-suite
Comment (Pepo48) [reply testsuite/integration-arquillian/tests/base/testsuites/fips-suite]: done as I mentioned above.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java
Comment (Pepo48) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java]: I simplified and changed the structure in order to improve readability. So right now we have the parent class `KeycloakVaultTest` with the common functionality for both `KeycloakKeystoreVaultTest` and `KeycloakKeystorePlaintextTest`. These two just start the server with necessary parameters while using a different provider (based on the `@EnableVault` annotation). All assertion is happening the parent class now.

I don't think that an abstraction is necessary - right now there is no need for overriding a method with specific functionality, the code is the same for both the children. WDYT?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java
Comment (Pepo48) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/vault/KeycloakVaultTest.java]: Do you think that this is still relevant? (see my previous comment)

**Comment by Pepo48** [comment]
@vmuzikar thanks for the review, I tried my best to address everything. I still expect some failures within the FIPS UT, since the bcfks is still required in the rest of the dist. I'm not sure whether we want to [skip it](https://github.com/keycloak/keycloak/pull/19644/commits/66e3d0c06f9e62a3e159b4dc27a34b91fc1f9e8d#diff-ebf02c48842e07db416b4773820dc5cd798b148da3f6915f7814a67a92d6c7d1L23) for now or not. Wdyt?

File: services/src/test/java/org/keycloak/vault/KeystoreVaultProviderTest.java
Comment (vmuzikar) [inline services/src/test/java/org/keycloak/vault/KeystoreVaultProviderTest.java:26]: I think this `assume` is not quite correct. Do we really want to run this test just in FIPS mode?

File: services/src/test/java/org/keycloak/vault/KeystoreVaultProviderTest.java
Comment (vmuzikar) [inline services/src/test/java/org/keycloak/vault/KeystoreVaultProviderTest.java:72]: Sorry, I missed this before. There should be something like this to ensure the test fails if there's no exception thrown.
```suggestion
        assertThrows("java.io.IOException: Invalid keystore format", RuntimeException.class, () -> provider.obtainSecret("alias"));
```

**Review by vmuzikar — COMMENTED** [review]
@Pepo48 Thanks for the update! Added two more comments.

On top of that, we'll need release notes but we can handle that as a follow-up and cover both KC Vault and SmallRye Keystore in a same note.

File: services/src/test/java/org/keycloak/vault/KeystoreVaultProviderTest.java
Comment (vmuzikar) [reply services/src/test/java/org/keycloak/vault/KeystoreVaultProviderTest.java:26]: Sorry, it's the other way around. I got confused :)

**Review by vmuzikar — DISMISSED** [review]
LGTM! Thanks @Pepo48, nice work

**Comment by vmuzikar** [comment]
j

**Review by abstractj — DISMISSED** [review]
Changes LGTM, but I would wait for @pedroigor approval considering his participation during the review of this PR.

File: docs/guides/server/vault.adoc
Comment (pedroigor) [inline docs/guides/server/vault.adoc:64]: Perhaps `keytool -importpass -alias <realm-name>_<alias> -keystore keystore.p12 -storepass keystorepassword`?

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/VaultKeystorePropertyMappers.java
Comment (pedroigor) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/VaultKeystorePropertyMappers.java:7]: Why a separate mapper type and not just re-use the `VaultPropertyMappers`? 

Same for keeping the file vault ...

**Review by pedroigor — APPROVED** [review]
@Pepo48 Looking forward to using the vault. Glad to see this finally happening :)

**Review by vmuzikar — APPROVED** [review]

**Pull request merged**
